File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> dr = {0, 0, 1, -1};
vector<int> dc = {1, -1, 0, 0};

int solution(int N, int M, vector<vector<int>> board) {
    vector<vector<bool>> visited(N, vector<bool>(M, false));
    int group = 0;
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < M; j++) {
            if (board[i][j] == 1 && !visited[i][j]) {
                visited[i][j] = true;
                stack<pair<int, int>> st;
                st.push({i, j});
                while (!st.empty()) {
                    auto [r, c] = st.top();
                    st.pop();

                    for (int idx = 0; idx < 4; idx++) {
                        int nr = r + dr[idx];
                        int nc = c + dc[idx];

                        if (0 <= nr && nr < N && 0 <= nc && nc < M &&
                            board[nr][nc] == 1 && !visited[nr][nc]) {
                            visited[nr][nc] = true;
                            st.push({nr, nc});
                        }
                    }
                }
                group++;
            }
        }
    }
    return group;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int M, N, K;
        cin >> M >> N >> K;

        // board[row][col] = board[y][x]
        vector<vector<int>> board(N, vector<int>(M, 0));

        for (int i = 0; i < K; i++) {
            int x, y;
            cin >> x >> y;
            board[y][x] = 1;
        }

        cout << solution(N, M, board) << '\n';
    }

    return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.py
import sys

input = sys.stdin.readline

d = [(0, 1), (0, -1), (1, 0), (-1, 0)]

T = int(input())

for _ in range(T):
    M, N, K = map(int, input().split())

    board = [[0] * M for _ in range(N)]

    for _ in range(K):
        x, y = map(int, input().split())
        board[y][x] = 1

    visited = [[False] * M for _ in range(N)]
    group = 0
    for i in range(N):
        for j in range(M):
            if board[i][j] and not visited[i][j]:
                visited[i][j] = True
                stack = [(i, j)]
                while stack:
                    r, c = stack.pop()

                    for dr, dc in d:
                        nr, nc = r + dr, c + dc
                        if (
                            0 <= nr < N
                            and 0 <= nc < M
                            and board[r][c]
                            and not visited[nr][nc]
                        ):
                            visited[nr][nc] = True
                            stack.append((nr, nc))

                group += 1

    print(group)
